### DISCUSSION TITLE
Question on How EntropyBottleneck's Parameters are Updated

### DISCUSSION THREAD

**Discussion by Will0x6c5f** [InterDigitalInc/CompressAI#177] [Q&A] [answered]
Hi there,
I'm currently working on this paper : [Variational image compression with a scale hyperprior](https://arxiv.org/abs/1802.01436) and find this implementation [(bmshj2018)](https://interdigitalinc.github.io/CompressAI/zoo.html#bmshj2018-hyperprior). It seems to me that this model use a MLP(or something like that) to approximate the 'rate' part of the rate-distortion loss according to appendix 6.1 and 6.2 in the original paper. 
However, I didn't find the corresponding loss to update this EntropyBottleneck's parameter (which should exist in order to optimize the matrices and biases inside it), i. e. **There is no measure on how well this surrogate network is approximating the true distribution**.
So how does this module update itself ? If the only loss affecting EntropyBottleneck is the 'rate' loss, how can the network ensure it will converge to the true distribution instead of returning all zero ? Is this related to the math property of that particular kind of "MLP"?  I wonder if my understanding is correct or I missed something in the paper. 

Thank you for your time

**Comment by YodaEmbedding** [comment, marked as answer]
`EntropyBottleneck.forward` computes the likelihoods that are used for the rate loss:
 https://github.com/InterDigitalInc/CompressAI/blob/677506b4b3f8923fcf60ef407b37082762c24b5d/compressai/entropy_models/entropy_models.py#L501

These likelihoods are computed using an input tensor and the matrices/biases:

https://github.com/InterDigitalInc/CompressAI/blob/677506b4b3f8923fcf60ef407b37082762c24b5d/compressai/entropy_models/entropy_models.py#L436-L469

The main "net" optimizer includes the matrices/biases in the list of parameters it optimizes. Therefore, these parameters are updated in such a way that they minimize the rate loss.

**Reply by Will0x6c5f** [reply]
Thank you for your reply !

But the problem remains... I'm aware that the matrices/biases are updated to minimize the rate loss. However, these matrices/biases are meant to approximate the **actual distribution** such that the _likelihood function take a value as input and return the approximate probability/likelihood of that value.
So the problem really is about how to ensure  these matrices/biases will produce the approximately correct likelihoods only minimizing the rate loss ? After all, this rate loss can easily converge to zero as long as all the matrices/biases are set to zero, but that will be no where around the true likelihood that we expect to get.

Please correct me if I missed something.

**Reply by YodaEmbedding** [reply]
The distortion term in `L = R + λD` ensures that `R` does not go to `0`. The `EntropyBottleneck` does not modify the input tensor data in any way, other than quantization noise. Thus, it has nearly no effect on `D`.

The rate `R` equals the cross-entropy `H(p; q)` between the true distribution `p` of the input tensors and the approximating distribution `q`. When `p = q`, the cross-entropy (and therefore, the rate) is minimized. If `q` is anything else, the rate loss is not at a minimum w.r.t. the target latent input distribution `p`. Since the optimizer minimizes the loss, it will do its best to ensure `q` tends towards `p` by appropriately adjusting the matrices/biases.

**Reply by Will0x6c5f** [reply]
I think I might get the rough idea of what's going on... but to ensure my understanding is correct, I'd like to confirm a few more things:
Generally R is an upper bound of the actual bit rate of that latent code... ? 
And only at the best situation with optimal matrices/biases this R will give the actual entropy of one latent code... ?

**Reply by YodaEmbedding** [reply]
`R` is an upper bound on rate because it is always possible that the input may be further compressed via some other distribution `q'` that is closer to `p`. When `p = q`, the current entropy model cannot get any better by adjusting `q` (which is a deterministic function of the matrices/biases).

Note that the `EntropyBottleneck` entropy model assumes that all symbols within the ith channel are i.i.d. and are drawn from the same distribution `q_i`. In reality, this assumption is not accurate since there are clear visible relationships between these variables when we visualize the feature maps. So, there exists a better entropy model for our data that is capable of giving a tighter upper bound on `R`. The ideal entropy model would allow us to compress to exactly the true entropy `R = H(y)`.

**Reply by Will0x6c5f** [reply]
Got it ! Thank you soooo much for answering my question ! ! 
Appreciate your sharing !
